mp4parse/src/lib.rs

```diff
@@ -1105,16 +1117,77 @@ fn read_flac_metadata<T: Read>(src: &mut BMFFBox<T>) -> Result<FLACMetadataBlock
     })
 }
 
-fn read_esds<T: Read>(src: &mut BMFFBox<T>) -> Result<AudioCodecSpecific> {
+fn read_esds<T: Read>(src: &mut BMFFBox<T>) -> Result<EsDescriptor> {
+    // Parsing DecoderConfig descriptor to get the codec type.
     let (_, _) = try!(read_fullbox_extra(src));
 
     let esds_size = src.head.size - src.head.offset - 4;
     if esds_size > BUF_SIZE_LIMIT {
         return Err(Error::InvalidData("esds box exceeds BUF_SIZE_LIMIT"));
     }
-    let esds = try!(read_buf(&mut src.content, esds_size as usize));
+    let esds_array = try!(read_buf(src, esds_size as usize));
+
+    // clone a esds cursor for parsing.
+    let esds = &mut Cursor::new(esds_array.clone());
```

Comment:
Rather than cloning the data, share it by creating the Cursor with a ref.  You just need to arrange for the Cursor to goe out of scope before you try to move esds_array into EsDescriptor.